core/types/block.go

```diff
@@ -231,9 +233,31 @@ func NewBlock(header *Header, txs []*Transaction, uncles []*Header, receipts []*
 	if len(txs) == 0 {
 		b.header.TxHash = EmptyTxsHash
 	} else {
-		b.header.TxHash = DeriveSha(Transactions(txs), hasher)
 		b.transactions = make(Transactions, len(txs))
+
+		// Collect Signatures
+		var signatures []bls.Signature
+		for _, tx := range txs {
+			if tx.Type() == BLSTxType {
+				sig, err := bls.SignatureFromBytes(tx.Signature())
+				if err != nil {
+					panic("could not convert bytes to BLS Signature")
+				}
+				signatures = append(signatures, sig)
+				// All transactions in the block will be added without the signature field set
+				tx.SetSignature(nil)
+			}
+		}
 		copy(b.transactions, txs)
+
+		// Aggregate BLS Signatures
+		var aggregatedSig []byte
+		if signatures != nil {
+			aggSig := bls.AggregateSignatures(signatures)
+			aggregatedSig = aggSig.Marshal()
+		}
+		b.header.AggregatedSig = aggregatedSig
+		b.header.TxHash = DeriveSha(Transactions(txs), hasher)
```

Comment:
Because I unset the signature, I believe this will cause a mismatch if the Header txHash was derived before unsetting, so I moved it to after the BLS txs got included in the block